#!/bin/bash -l
#SBATCH -J StringTie_1
#SBATCH -p core
#SBATCH -n 1
#SBATCH -t 04:00:00
#SBATCH -A snic2019-3-35
#SBATCH -M snowy
#SBATCH --mail-user veronika.mrazek@ebc.uu.se
#SBATCH --mail-type=ALL

# load modules
module load bioinfo-tools
module load StringTie/1.3.3





## PATHS and filenames

#initial folder
SRCDIR=$(pwd)

#read name of BAM file (with path)
BAM1=${1:?msg}

#read name of BAM file (without path)
rawBAM1=${2:?msg}

echo $BAM1

# output folder (root name)
RR=${3:?msg}

#sample name
SNAME=${4:?msg}

#Annotation file
ANNF=${5:?msg}

echo $ANNF
echo






# Run StringTie

stringtie \
      $BAM1 \
      -p 1 \
      -G $ANNF \
      -o $RR/${SNAME}.STRINGTIE-out.assembled-transcripts.gtf \
      -A $RR/${SNAME}.STRINGTIE-out.gene_abund.tab \
      -C $RR/${SNAME}.STRINGTIE-out.cov_refs.gtf \
          -e







#Notes (StringTie)
# -o    output path/file name for the assembled transcripts GTF (default: stdout)
# -p    number of threads (CPUs) to use (default: 1)
# -m    minimum assembled transcript length (default: 200)
# -A <gene_abund.tab>   Gene abundances will be reported (tab delimited format) in the output file with the given name.
# -G <ref_ann.gff>      Use the reference annotation file (in GTF or GFF3 format) to guide the assembly process. The output will include expressed reference transcripts as well as any novel transcripts that are assembled. This option is required by options -B, -b, -e, -C (see below).
# -C <cov_refs.gtf>     StringTie outputs a file with the given name with all transcripts in the provided reference file that are fully covered by reads (requires -G).

# -e    Limits the processing of read alignments to only estimate and output the assembled transcripts matching the reference transcripts given with the -G option (requires -G,          recommended for -B/-b). With this option, read bundles with no reference transcripts will be entirely skipped, which may provide a considerable speed boost when the given        set of reference transcripts is limited to a set of target genes, for example.
#               The -e option is not required but recommended for this run in order to produce more accurate abundance estimations of the input transcripts (from StringTie manual)
#               This option is required when preparing files generated by StringTie for use with DESeq2 and edgeR.




#note: runtime:   10min     (1 cores);




