---
title: "RNA_hostplant_topGO"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown



```{r RNA_topGO_test}
#obs!!
#rm(list=ls())  #run it only in console before running the script, should not be used in Rmarkdown

library(topGO)
library(AnnotationDbi)
library(GO.db)
library(org.Dm.eg.db)

sessionInfo()

##################################
#get the data
#create gene list from list created with Stringtie where the gene alias in orthologs to D. melanogaster is included. 
#create a background set of genes from the specific dataset filtered to include the genes with basemean > 10
#select genes that are differentially expresssed from the different DE-datasets 

#first create gene list with geneID this can be reused for all datasets
geneList_dmel=read.table("gene_list_sin_dmel_stringtie.txt", header = T)
#get rid of the "-" (unannotated genes) 
geneList_dmel_filtered <- subset(geneList_dmel, !geneList_dmel$GeneName== "-")
#length(geneList_dmel_filtered$GeneID)
#[1] 10926

#background genes with only D.mel names
#bg_genes<-as.character(geneList_dmel_filtered[,2])


#Background genes for instar 3, span, filtered to over 10 basemean
bg_instar3span <- read.table("../../Results/4_DEseq2/06_Spa_Hostplant_effect/Filtered_base_Mean_10_Instar3_Spa_Host.txt", header = T)
bg_instar3span$GeneID <- rownames(bg_instar3span)
length(bg_instar3span$GeneID)
#[1] 6709

#DE genes in Cat instar3
candidateGenes_instar3span <- read.table("../../Results/4_DEseq2/06_Spa_Hostplant_effect/Filtered_base_Mean_10_LFC_1.0_P0.05_Instar3_Spa_Host.txt", header = T)
candidateGenes_instar3span$GeneID <- rownames(candidateGenes_instar3span)
length(candidateGenes_instar3span$GeneID)

#get the D mel gene names
bg_instar3span_Dmel <- merge(bg_instar3span, geneList_dmel_filtered, by.x="GeneID", by.y="GeneID")
canditateGenes_instar3span_Dmel <- merge(candidateGenes_instar3span, geneList_dmel_filtered, by.x="GeneID", by.y="GeneID")

length(canditateGenes_instar3span_Dmel$GeneID)
#[1] 33

candidateGenes_instar3span_list <- as.character(canditateGenes_instar3span_Dmel$GeneName)
length(candidateGenes_instar3span_list) #33 genes
bg_instar3span_Dmel_list <- as.character(bg_instar3span_Dmel$GeneName)
length(bg_instar3span_Dmel_list) 

###########################################
#GO analysis
# GeneName to GO mapping, get nr of GO terms per gene in the candadate genelist
GOtoalias_instar3span_BP <- annFUN.org("BP", feasibleGenes = candidateGenes_instar3span_list, mapping = "org.Dm.eg.db", ID = "Alias")
GeneIDtoGO_instar3span_BP <- inverseList(GOtoalias_instar3span_BP) # only 6 genes with GO annotations
GeneIDtoGO_instar3span_BP

GOtoalias_instar3span_MF <- annFUN.org("MF", feasibleGenes = candidateGenes_instar3span_list, mapping = "org.Dm.eg.db", ID = "Alias")
GeneIDtoGO_instar3span_MF <- inverseList(GOtoalias_instar3span_MF) # only 6 genes with GO annotations
GeneIDtoGO_instar3span_MF

GOtoalias_instar3span_CC <- annFUN.org("CC", feasibleGenes = candidateGenes_instar3span_list, mapping = "org.Dm.eg.db", ID = "Alias")
GeneIDtoGO_instar3span_CC <- inverseList(GOtoalias_instar3span_CC) # only 6 genes with GO annotations
GeneIDtoGO_instar3span_CC

#convert these list into a dataframe to be able to merge it later in the result folder, add a column so that genealiases ca be a separate column
GeneIDtoGO_instar3span_BP.df <- do.call(rbind, lapply(GeneIDtoGO_instar3span_BP, as.data.frame))
colnames(GeneIDtoGO_instar3span_BP.df) <- c("BP")
GeneIDtoGO_instar3span_BP.df$GeneName <- rownames(GeneIDtoGO_instar3span_BP.df)

# GeneIDtoGO_instar3span_MF.df <- do.call(rbind, lapply(GeneIDtoGO_instar3span_MF, as.data.frame))
# colnames(GeneIDtoGO_instar3span_MF.df) <- c("MF")
# GeneIDtoGO_instar3span_MF.df$GeneName <- rownames(GeneIDtoGO_instar3span_MF.df)
# 
# GeneIDtoGO_instar3span_CC.df <- do.call(rbind, lapply(GeneIDtoGO_instar3span_CC, as.data.frame))
# colnames(GeneIDtoGO_instar3span_CC.df) <- c("CC")
# GeneIDtoGO_instar3span_CC.df$GeneName <- rownames(GeneIDtoGO_instar3span_CC.df)

#creating an topGOdata-object
#first give the candidate genes a "score" Listed as 1 and the other are listed as 0
geneList_instar3span <- factor(as.integer(bg_instar3span_Dmel_list %in% candidateGenes_instar3span_list))
names(geneList_instar3span) <- bg_instar3span_Dmel_list
str(geneList_instar3span)

sampleGOdata_instar3span_BP <- new(Class = "topGOdata", description="instar3span_hostplant", ontology="BP", allGenes=geneList_instar3span, annot=annFUN.org, mapping = "org.Dm.eg.db", ID = "Alias")
#get a short summary
sampleGOdata_instar3span_BP

# sampleGOdata_instar3span_MF <- new(Class = "topGOdata", description="instar3span_hostplant", ontology="MF", allGenes=geneList_instar3span, annot=annFUN.org, mapping = "org.Dm.eg.db", ID = "Alias")
# #get a short summary
# sampleGOdata_instar3span_MF
# 
# sampleGOdata_instar3span_CC <- new(Class = "topGOdata", description="instar3span_hostplant", ontology="CC", allGenes=geneList_instar3span, annot=annFUN.org, mapping = "org.Dm.eg.db", ID = "Alias")
# #get a short summary
# sampleGOdata_instar3span_CC




###########################################################
#run a significance test 
#The elim and weight algorithms were introduced in Alexa et al. (2006). The default algorithm used by the topGO package is a mixture between #the elim and the weight algorithms and it will be referred as weight01.
weight_fisher_result_instar3span_BP=runTest(sampleGOdata_instar3span_BP, algorithm='weight01', statistic='fisher')
#get the result summary
weight_fisher_result_instar3span_BP
#get the p-values with score
#weight_fisher_result_pvalues_instar3span <- score(weight_fisher_result_instar3span)
#more results
#geneData(weight_fisher_result_instar3span_BP)
#Write results in table
results_table_instar3span_BP <- GenTable(object = sampleGOdata_instar3span_BP, weight=weight_fisher_result_instar3span_BP, orderBy="weight", topNodes=20)
#results_table_instar3span_BP

#get only GOterms sign enriched , sorted by GO term
#sort(subset(results_table_instar3span, results_table_instar3span$weight <= 0.05)[ ,1])
#get only significantly enriched GOterm with result table
results_table_instar3span_sign_BP <- subset(results_table_instar3span_BP, results_table_instar3span_BP$weight <= 0.05)
results_table_instar3span_sign_BP

#checking the distribution of annotations
#results_table_instar3span_long <- GenTable(object = sampleGOdata_instar3span_BP, weight=weight_fisher_result_instar3span, orderBy="weight", topNodes=1000)
#hist(results_table_instar3span_long$Annotated, breaks = 500, xlim = c(0,100))

##########################################
#collecting the results in one table
#GeneID and GOterms assoc with those genes, GeneIDtoGO_instar3span_BP etc, merging _BP, MF and CC to one table?
#GOterms and p-values (table), results_table_instar3span_sign_BP

result_table_GO_BP <- merge(results_table_instar3span_sign_BP, GeneIDtoGO_instar3span_BP.df, by.x="GO.ID", by.y="BP")
#remove the extra suffix ork1.1 etc
result_table_GO_BP$GeneName=gsub(pattern = "\\.[0-99]*$",  replacement = "", x = result_table_GO_BP$GeneName, fixed = FALSE)
#result_table_GO_BP


#merge df to one list with GO.ID, GOterms, gene-alias and gene names, and p-value for enrichment-test
#get genenames from the org....db, can be resused in the different datasets
GeneListNames_db <- select(org.Dm.eg.db, keys=keys(org.Dm.eg.db), columns=c("GENENAME", "ALIAS"))
result_table_GO_BP_temp <- merge(result_table_GO_BP, GeneListNames_db, by.x="GeneName", by.y="ALIAS")
#to add a column with GeneID for L.sinapis
result_table_GO_BP_genenames <- merge(result_table_GO_BP_temp, geneList_dmel_filtered, by.x="GeneName", by.y="GeneName")
result_table_GO_BP_genenames

#write result to table
write.csv(result_table_GO_BP_genenames, file="../../Results/6_topGO/result_GO_instar3span_hostplant_BP.table")



#gives all genes with the sign GOs, but also genes without sign DE
#genesInTerm(sampleGOdata_instar3span_BP, results_table_instar3span_sign$GO.ID)

#To get a table of all genes with the sign G0.id, not necessary
# myterms=results_table_instar3span_sign$GO.ID
# mygenes=genesInTerm(sampleGOdata_instar3span_BP, results_table_instar3span_sign$GO.ID)
# 
# var=c()
# # for (i in 1:length(myterms))
# + {
# +     myterm=myterms[i]
# +     mygenesforterm= mygenes[myterm][[1]]
# +     mygenesforterm=paste(mygenesforterm, collapse=',')
# +     var[i]=paste("GOTerm",myterm,"genes-",mygenesforterm)
# + }
#  


```

